File: snippets/screens/swiftui/chat_screen/2/chat_screen.swift
import SwiftUI

struct ChatScreen: View {
    let username: String
    @StateObject private var viewModel = ChatViewModel()
    @State private var messageText = ""
    @FocusState private var isMessageFieldFocused: Bool
    @State private var showingAttachmentOptions = false

    var body: some View {
        VStack(spacing: 0) {
            // Chat messages
            ScrollViewReader { proxy in
                ScrollView {
                    LazyVStack(spacing: 12) {
                        ForEach(viewModel.messages) { message in
                            MessageBubble(message: message)
                                .id(message.id)
                        }

                        if viewModel.isTyping {
                            TypingIndicator()
                                .id("typing")
                        }
                    }
                    .padding(.horizontal, 16)
                    .padding(.vertical, 12)
                }
                .onAppear {
                    scrollToBottom(proxy: proxy)
                }
                .onChange(of: viewModel.messages.count) { _ in
                    scrollToBottom(proxy: proxy)
                }
            }

            // Message input
            MessageInputBar(
                text: $messageText,
                isFocused: $isMessageFieldFocused,
                onSend: {
                    viewModel.sendMessage(messageText)
                    messageText = ""
                },
                onAttachment: {
                    showingAttachmentOptions = true
                },
                onVoiceRecord: {
                    viewModel.startVoiceRecording()
                }
            )
        }
        .navigationTitle(username)
        .navigationBarTitleDisplayMode(.inline)
        .toolbar {
            ToolbarItem(placement: .principal) {
                VStack(spacing: 2) {
                    Text(username)
                        .font(.headline)
                        .fontWeight(.semibold)

                    Text(viewModel.isOnline ? "Online" : "Last seen \(viewModel.lastSeen)")
                        .font(.caption)
                        .foregroundColor(viewModel.isOnline ? .green : .secondary)
                }
            }

            ToolbarItemGroup(placement: .navigationBarTrailing) {
                Button(action: { viewModel.startVideoCall() }) {
                    Image(systemName: "video")
                }

                Button(action: { viewModel.startVoiceCall() }) {
                    Image(systemName: "phone")
                }

                Menu {
                    Button("View Profile") { viewModel.viewProfile() }
                    Button("Mute") { viewModel.muteChat() }
                    Button("Block") { viewModel.blockUser() }
                    Button("Report") { viewModel.reportUser() }
                    Button("Clear Chat") { viewModel.clearChat() }
                } label: {
                    Image(systemName: "ellipsis")
                }
            }
        }
        .onAppear {
            viewModel.loadMessages(for: username)
        }
        .actionSheet(isPresented: $showingAttachmentOptions) {
            ActionSheet(
                title: Text("Send"),
                buttons: [
                    .default(Text("Camera")) { viewModel.sendMedia(.camera) },
                    .default(Text("Photo Library")) { viewModel.sendMedia(.gallery) },
                    .default(Text("Document")) { viewModel.sendMedia(.document) },
                    .default(Text("Location")) { viewModel.sendLocation() },
                    .default(Text("Contact")) { viewModel.sendContact() },
                    .cancel()
                ]
            )
        }
    }

    private func scrollToBottom(proxy: ScrollViewProxy) {
        if let lastMessage = viewModel.messages.last {
            withAnimation(.easeOut(duration: 0.3)) {
                proxy.scrollTo(lastMessage.id, anchor: .bottom)
            }
        }
    }
}

struct MessageBubble: View {
    let message: ChatMessage

    private var isCurrentUser: Bool {
        message.senderId == "current_user"
    }

    var body: some View {
        HStack {
            if !isCurrentUser {
                AsyncImage(url: URL(string: "https://example.com/avatar3.jpg")) { image in
                    image
                        .resizable()
                        .aspectRatio(contentMode: .fill)
                } placeholder: {
                    Circle()
                        .fill(Color.gray.opacity(0.3))
                }
                .frame(width: 24, height: 24)
                .clipShape(Circle())
            }

            if isCurrentUser {
                Spacer()
            }

            VStack(alignment: isCurrentUser ? .trailing : .leading, spacing: 4) {
                MessageContent(message: message, isCurrentUser: isCurrentUser)

                HStack(spacing: 4) {
                    Text(formatTime(message.timestamp))
                        .font(.caption2)
                        .foregroundColor(.secondary)

                    if isCurrentUser {
                        Image(systemName: message.isRead ? "checkmark.circle.fill" : "checkmark.circle")
                            .font(.caption2)
                            .foregroundColor(message.isRead ? .blue : .secondary)
                    }
                }
            }

            if !isCurrentUser {
                Spacer()
            }

            if isCurrentUser {
                AsyncImage(url: URL(string: "https://example.com/current_user_avatar.jpg")) { image in
                    image
                        .resizable()
                        .aspectRatio(contentMode: .fill)
                } placeholder: {
                    Circle()
                        .fill(Color.gray.opacity(0.3))
                }
                .frame(width: 24, height: 24)
                .clipShape(Circle())
            }
        }
    }

    private func formatTime(_ date: Date) -> String {
        let formatter = DateFormatter()
        if Calendar.current.isDateInToday(date) {
            formatter.timeStyle = .short
        } else {
            formatter.dateStyle = .short
        }
        return formatter.string(from: date)
    }
}

struct MessageContent: View {
    let message: ChatMessage
    let isCurrentUser: Bool

    var body: some View {
        Group {
            switch message.type {
            case .text:
                if let text = message.text {
                    Text(text)
                        .font(.body)
                        .foregroundColor(isCurrentUser ? .white : .primary)
                        .padding(.horizontal, 16)
                        .padding(.vertical, 10)
                        .background(isCurrentUser ? Color.blue : Color.gray.opacity(0.2))
                        .clipShape(RoundedRectangle(cornerRadius: 20))
                }

            case .image:
                if let imageUrl = message.imageUrl {
                    AsyncImage(url: URL(string: imageUrl)) { image in
                        image
                            .resizable()
                            .aspectRatio(contentMode: .fill)
                    } placeholder: {
                        Rectangle()
                            .fill(Color.gray.opacity(0.3))
                            .overlay(
                                Image(systemName: "photo")
                                    .foregroundColor(.gray)
                            )
                    }
                    .frame(width: 200, height: 200)
                    .clipShape(RoundedRectangle(cornerRadius: 12))
                    .onTapGesture {
                        // Show full screen image
                    }
                }

            case .voice:
                VoiceMessageView(isCurrentUser: isCurrentUser)

            case .video:
                VideoMessageView(isCurrentUser: isCurrentUser)
            }
        }
    }
}

struct VoiceMessageView: View {
    let isCurrentUser: Bool
    @State private var isPlaying = false

    var body: some View {
        HStack {
            Button(action: { isPlaying.toggle() }) {
                Image(systemName: isPlaying ? "pause.fill" : "play.fill")
                    .foregroundColor(isCurrentUser ? .white : .blue)
            }

            WaveformView()
                .frame(height: 20)

            Text("0:15")
                .font(.caption)
                .foregroundColor(isCurrentUser ? .white : .secondary)
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 10)
        .background(isCurrentUser ? Color.blue : Color.gray.opacity(0.2))
        .clipShape(RoundedRectangle(cornerRadius: 20))
        .frame(width: 200)
    }
}

struct VideoMessageView: View {
    let isCurrentUser: Bool

    var body: some View {
        Rectangle()
            .fill(Color.gray.opacity(0.3))
            .frame(width: 200, height: 150)
            .overlay(
                Button(action: {
                    // Play video
                }) {
                    Image(systemName: "play.circle.fill")
                        .font(.largeTitle)
                        .foregroundColor(.white)
                }
            )
            .clipShape(RoundedRectangle(cornerRadius: 12))
    }
}

struct WaveformView: View {
    var body: some View {
        HStack(spacing: 2) {
            ForEach(0..<20, id: \.self) { _ in
                Rectangle()
                    .fill(Color.blue.opacity(0.6))
                    .frame(width: 2, height: CGFloat.random(in: 4...16))
            }
        }
    }
}

struct TypingIndicator: View {
    @State private var animating = false

    var body: some View {
        HStack {
            AsyncImage(url: URL(string: "https://example.com/avatar3.jpg")) { image in
                image
                    .resizable()
                    .aspectRatio(contentMode: .fill)
            } placeholder: {
                Circle()
                    .fill(Color.gray.opacity(0.3))
            }
            .frame(width: 24, height: 24)
            .clipShape(Circle())

            HStack(spacing: 4) {
                ForEach(0..<3) { index in
                    Circle()
                        .fill(Color.gray)
                        .frame(width: 8, height: 8)
                        .scaleEffect(animating ? 1.2 : 0.8)
                        .animation(
                            Animation.easeInOut(duration: 0.6)
                                .repeatForever()
                                .delay(Double(index) * 0.2),
                            value: animating
                        )
                }
            }
            .padding(.horizontal, 16)
            .padding(.vertical, 10)
            .background(Color.gray.opacity(0.2))
            .clipShape(RoundedRectangle(cornerRadius: 20))

            Spacer()
        }
        .onAppear {
            animating = true
        }
    }
}

struct MessageInputBar: View {
    @Binding var text: String
    var isFocused: FocusState<Bool>.Binding
    let onSend: () -> Void
    let onAttachment: () -> Void
    let onVoiceRecord: () -> Void

    var body: some View {
        VStack {
            Divider()

            HStack(spacing: 12) {
                Button(action: onAttachment) {
                    Image(systemName: "plus")
                        .font(.title2)
                        .foregroundColor(.blue)
                }

                TextField("Type a message...", text: $text)
                    .focused(isFocused)
                    .textFieldStyle(.roundedBorder)

                Button(action: text.isEmpty ? onVoiceRecord : onSend) {
                    Image(systemName: text.isEmpty ? "mic.fill" : "paperplane.fill")
                        .font(.title2)
                        .foregroundColor(.blue)
                }
            }
            .padding(.horizontal, 16)
            .padding(.vertical, 8)
        }
    }
}

@MainActor
class ChatViewModel: ObservableObject {
    @Published var messages: [ChatMessage] = []
    @Published var isTyping = false
    @Published var isOnline = true
    @Published var lastSeen = "2 minutes ago"

    func loadMessages(for username: String) {
        messages = [
            ChatMessage(
                id: "1",
                senderId: "alice_wonder",
                text: "Hey! How are you?",
                timestamp: Date().addingTimeInterval(-7200),
                type: .text,
                isRead: true
            ),
            ChatMessage(
                id: "2",
                senderId: "current_user",
                text: "I'm good! Just working on some projects. How about you?",
                timestamp: Date().addingTimeInterval(-7100),
                type: .text,
                isRead: true
            ),
            ChatMessage(
                id: "3",
                senderId: "alice_wonder",
                imageUrl: "https://example.com/shared_image.jpg",
                timestamp: Date().addingTimeInterval(-3600),
                type: .image,
                isRead: true
            ),
            ChatMessage(
                id: "4",
                senderId: "alice_wonder",
                text: "Check out this cool photo I took!",
                timestamp: Date().addingTimeInterval(-3540),
                type: .text,
                isRead: true
            ),
            ChatMessage(
                id: "5",
                senderId: "current_user",
                text: "Wow, that's amazing! 😍",
                timestamp: Date().addingTimeInterval(-1800),
                type: .text,
                isRead: true
            ),
            ChatMessage(
                id: "6",
                senderId: "alice_wonder",
                text: "Thanks! Want to meet up this weekend?",
                timestamp: Date().addingTimeInterval(-300),
                type: .text,
                isRead: false
            )
        ]

        // Simulate typing
        DispatchQueue.main.asyncAfter(deadline: .now() + 3) {
            self.isTyping = true

            DispatchQueue.main.asyncAfter(deadline: .now() + 2) {
                self.isTyping = false
            }
        }
    }

    func sendMessage(_ text: String) {
        guard !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return }

        let newMessage = ChatMessage(
            id: UUID().uuidString,
            senderId: "current_user",
            text: text,
            timestamp: Date(),
            type: .text,
            isRead: false
        )

        messages.append(newMessage)
    }

    func sendMedia(_ type: MediaType) {
        let newMessage = ChatMessage(
            id: UUID().uuidString,
            senderId: "current_user",
            imageUrl: "https://example.com/sent_media.jpg",
            timestamp: Date(),
            type: .image,
            isRead: false
        )

        messages.append(newMessage)
    }

    func sendLocation() {
        // Send location message
    }

    func sendContact() {
        // Send contact message
    }

    func startVoiceRecording() {
        let voiceMessage = ChatMessage(
            id: UUID().uuidString,
            senderId: "current_user",
            timestamp: Date(),
            type: .voice,
            isRead: false
        )

        messages.append(voiceMessage)
    }

    func startVideoCall() {
        // Start video call
    }

    func startVoiceCall() {
        // Start voice call
    }

    func viewProfile() {
        // Navigate to profile
    }

    func muteChat() {
        // Mute chat
    }

    func blockUser() {
        // Block user
    }

    func reportUser() {
        // Report user
    }

    func clearChat() {
        // Clear chat
        messages.removeAll()
    }
}

enum MediaType {
    case camera, gallery, document
}

struct ChatMessage: Identifiable {
    let id: String
    let senderId: String
    let text: String?
    let imageUrl: String?
    let timestamp: Date
    let type: ChatMessageType
    var isRead: Bool

    init(id: String, senderId: String, text: String? = nil, imageUrl: String? = nil, timestamp: Date, type: ChatMessageType, isRead: Bool) {
        self.id = id
        self.senderId = senderId
        self.text = text
        self.imageUrl = imageUrl
        self.timestamp = timestamp
        self.type = type
        self.isRead = isRead
    }
}

enum ChatMessageType {
    case text, image, voice, video
}

#Preview {
    NavigationView {
        ChatScreen(username: "alice_wonder")
    }
}

File: snippets/screens/flutter/chat_screen/1/chat_screen.dart
import 'package:flutter/material.dart';

class ChatScreen extends StatefulWidget {
  final String username;

  const ChatScreen({Key? key, required this.username}) : super(key: key);

  @override
  _ChatScreenState createState() => _ChatScreenState();
}

class _ChatScreenState extends State<ChatScreen>
    with SingleTickerProviderStateMixin {
  final TextEditingController _messageController = TextEditingController();
  final ScrollController _scrollController = ScrollController();

  late AnimationController _animationController;

  List<Message> messages = [
    Message(
      id: '1',
      senderId: 'alice_wonder',
      text: 'Hey! How are you?',
      timestamp: DateTime.now().subtract(Duration(hours: 2)),
      type: MessageType.text,
      isRead: true,
    ),
    Message(
      id: '2',
      senderId: 'current_user',
      text: 'I\'m good! Just working on some projects. How about you?',
      timestamp: DateTime.now().subtract(Duration(hours: 2, minutes: -5)),
      type: MessageType.text,
      isRead: true,
    ),
    Message(
      id: '3',
      senderId: 'alice_wonder',
      imageUrl: 'https://example.com/shared_image.jpg',
      timestamp: DateTime.now().subtract(Duration(hours: 1)),
      type: MessageType.image,
      isRead: true,
    ),
    Message(
      id: '4',
      senderId: 'alice_wonder',
      text: 'Check out this cool photo I took!',
      timestamp: DateTime.now().subtract(Duration(hours: 1, minutes: -1)),
      type: MessageType.text,
      isRead: true,
    ),
    Message(
      id: '5',
      senderId: 'current_user',
      text: 'Wow, that\'s amazing! 😍',
      timestamp: DateTime.now().subtract(Duration(minutes: 30)),
      type: MessageType.text,
      isRead: true,
    ),
    Message(
      id: '6',
      senderId: 'alice_wonder',
      text: 'Thanks! Want to meet up this weekend?',
      timestamp: DateTime.now().subtract(Duration(minutes: 5)),
      type: MessageType.text,
      isRead: false,
    ),
  ];

  bool isOnline = true;
  String lastSeen = '2 minutes ago';
  bool isTyping = false;
  String recordingPath = '';
  bool isRecording = false;

  @override
  void initState() {
    super.initState();
    _animationController = AnimationController(
      duration: Duration(milliseconds: 300),
      vsync: this,
    );

    // Simulate typing indicator
    _simulateTyping();

    // Auto scroll to bottom
    WidgetsBinding.instance.addPostFrameCallback((_) {
      _scrollToBottom();
    });
  }

  void _simulateTyping() {
    Future.delayed(Duration(seconds: 3), () {
      if (mounted) {
        setState(() {
          isTyping = true;
        });

        Future.delayed(Duration(seconds: 2), () {
          if (mounted) {
            setState(() {
              isTyping = false;
            });
          }
        });
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: _buildAppBar(),
      body: Column(
        children: [
          Expanded(
            child: ListView.builder(
              controller: _scrollController,
              padding: EdgeInsets.symmetric(vertical: 8),
              itemCount: messages.length + (isTyping ? 1 : 0),
              itemBuilder: (context, index) {
                if (isTyping && index == messages.length) {
                  return _buildTypingIndicator();
                }
                return _buildMessageBubble(messages[index]);
              },
            ),
          ),
          _buildMessageInput(),
        ],
      ),
    );
  }

  PreferredSizeWidget _buildAppBar() {
    return AppBar(
      backgroundColor: Colors.white,
      foregroundColor: Colors.black,
      elevation: 1,
      titleSpacing: 0,
      title: GestureDetector(
        onTap: _navigateToProfile,
        child: Row(
          children: [
            CircleAvatar(
              radius: 20,
              backgroundImage: NetworkImage('https://example.com/avatar3.jpg'),
            ),
            SizedBox(width: 12),
            Expanded(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text(
                    widget.username,
                    style: TextStyle(
                      fontWeight: FontWeight.bold,
                      fontSize: 16,
                    ),
                  ),
                  Text(
                    isOnline ? 'Online' : 'Last seen $lastSeen',
                    style: TextStyle(
                      fontSize: 12,
                      color: isOnline ? Colors.green : Colors.grey,
                    ),
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
      actions: [
        IconButton(
          icon: Icon(Icons.videocam),
          onPressed: _startVideoCall,
        ),
        IconButton(
          icon: Icon(Icons.call),
          onPressed: _startVoiceCall,
        ),
        PopupMenuButton<String>(
          onSelected: _handleMenuAction,
          itemBuilder: (context) => [
            PopupMenuItem(value: 'view_profile', child: Text('View Profile')),
            PopupMenuItem(value: 'mute', child: Text('Mute')),
            PopupMenuItem(value: 'block', child: Text('Block')),
            PopupMenuItem(value: 'report', child: Text('Report')),
            PopupMenuItem(value: 'clear_chat', child: Text('Clear Chat')),
          ],
        ),
      ],
    );
  }

  Widget _buildMessageBubble(Message message) {
    bool isMe = message.senderId == 'current_user';

    return Container(
      margin: EdgeInsets.symmetric(horizontal: 16, vertical: 2),
      child: Row(
        mainAxisAlignment: isMe ? MainAxisAlignment.end : MainAxisAlignment.start,
        crossAxisAlignment: CrossAxisAlignment.end,
        children: [
          if (!isMe) ...[
            CircleAvatar(
              radius: 12,
              backgroundImage: NetworkImage('https://example.com/avatar3.jpg'),
            ),
            SizedBox(width: 8),
          ],
          Flexible(
            child: Container(
              constraints: BoxConstraints(
                maxWidth: MediaQuery.of(context).size.width * 0.7,
              ),
              padding: EdgeInsets.symmetric(horizontal: 16, vertical: 10),
              decoration: BoxDecoration(
                color: isMe
                    ? Theme.of(context).primaryColor
                    : Colors.grey[200],
                borderRadius: BorderRadius.circular(20),
              ),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  if (message.type == MessageType.image)
                    _buildImageMessage(message),
                  if (message.type == MessageType.voice)
                    _buildVoiceMessage(message),
                  if (message.type == MessageType.text && message.text != null)
                    Text(
                      message.text!,
                      style: TextStyle(
                        color: isMe ? Colors.white : Colors.black,
                        fontSize: 16,
                      ),
                    ),
                  SizedBox(height: 4),
                  Row(
                    mainAxisSize: MainAxisSize.min,
                    children: [
                      Text(
                        _formatTime(message.timestamp),
                        style: TextStyle(
                          color: isMe ? Colors.white70 : Colors.grey[600],
                          fontSize: 12,
                        ),
                      ),
                      if (isMe) ...[
                        SizedBox(width: 4),
                        Icon(
                          message.isRead ? Icons.done_all : Icons.done,
                          size: 16,
                          color: message.isRead ? Colors.blue : Colors.white70,
                        ),
                      ],
                    ],
                  ),
                ],
              ),
            ),
          ),
          if (isMe) ...[
            SizedBox(width: 8),
            CircleAvatar(
              radius: 12,
              backgroundImage: NetworkImage('https://example.com/current_user_avatar.jpg'),
            ),
          ],
        ],
      ),
    );
  }

  Widget _buildImageMessage(Message message) {
    return GestureDetector(
      onTap: () => _viewFullScreenImage(message.imageUrl!),
      child: Container(
        width: 200,
        height: 200,
        margin: EdgeInsets.only(bottom: 8),
        decoration: BoxDecoration(
          borderRadius: BorderRadius.circular(12),
          image: DecorationImage(
            image: NetworkImage(message.imageUrl!),
            fit: BoxFit.cover,
          ),
        ),
        child: Container(
          decoration: BoxDecoration(
            borderRadius: BorderRadius.circular(12),
            gradient: LinearGradient(
              begin: Alignment.topCenter,
              end: Alignment.bottomCenter,
              colors: [Colors.transparent, Colors.black26],
            ),
          ),
        ),
      ),
    );
  }

  Widget _buildVoiceMessage(Message message) {
    return Container(
      width: 200,
      child: Row(
        children: [
          IconButton(
            icon: Icon(
              Icons.play_arrow,
              color: message.senderId == 'current_user' ? Colors.white : Colors.black,
            ),
            onPressed: () => _playVoiceMessage(message),
          ),
          Expanded(
            child: Container(
              height: 4,
              decoration: BoxDecoration(
                color: message.senderId == 'current_user'
                    ? Colors.white30
                    : Colors.grey[400],
                borderRadius: BorderRadius.circular(2),
              ),
              child: LinearProgressIndicator(
                value: 0.3, // Sample progress
                backgroundColor: Colors.transparent,
                valueColor: AlwaysStoppedAnimation(
                  message.senderId == 'current_user' ? Colors.white : Colors.blue,
                ),
              ),
            ),
          ),
          SizedBox(width: 8),
          Text(
            '0:15',
            style: TextStyle(
              color: message.senderId == 'current_user' ? Colors.white70 : Colors.grey[600],
              fontSize: 12,
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildTypingIndicator() {
    return Container(
      margin: EdgeInsets.symmetric(horizontal: 16, vertical: 8),
      child: Row(
        children: [
          CircleAvatar(
            radius: 12,
            backgroundImage: NetworkImage('https://example.com/avatar3.jpg'),
          ),
          SizedBox(width: 8),
          Container(
            padding: EdgeInsets.symmetric(horizontal: 16, vertical: 10),
            decoration: BoxDecoration(
              color: Colors.grey[200],
              borderRadius: BorderRadius.circular(20),
            ),
            child: Row(
              mainAxisSize: MainAxisSize.min,
              children: [
                _buildTypingDot(0),
                SizedBox(width: 4),
                _buildTypingDot(1),
                SizedBox(width: 4),
                _buildTypingDot(2),
              ],
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildTypingDot(int index) {
    return AnimatedBuilder(
      animation: _animationController,
      builder: (context, child) {
        double value = (_animationController.value + index * 0.2) % 1.0;
        double opacity = (0.5 + 0.5 * (1.0 - (value * 2 - 1).abs())).clamp(0.0, 1.0);

        return Container(
          width: 8,
          height: 8,
          decoration: BoxDecoration(
            color: Colors.grey.withOpacity(opacity),
            shape: BoxShape.circle,
          ),
        );
      },
    );
  }

  Widget _buildMessageInput() {
    return Container(
      padding: EdgeInsets.all(16),
      decoration: BoxDecoration(
        color: Colors.white,
        border: Border(top: BorderSide(color: Colors.grey[200]!)),
      ),
      child: Row(
        children: [
          IconButton(
            icon: Icon(Icons.add, color: Theme.of(context).primaryColor),
            onPressed: _showAttachmentOptions,
          ),
          Expanded(
            child: TextField(
              controller: _messageController,
              decoration: InputDecoration(
                hintText: 'Type a message...',
                border: OutlineInputBorder(
                  borderRadius: BorderRadius.circular(25),
                  borderSide: BorderSide(color: Colors.grey[300]!),
                ),
                contentPadding: EdgeInsets.symmetric(horizontal: 16, vertical: 8),
              ),
              onChanged: (text) {
                // Handle typing indicator
              },
            ),
          ),
          SizedBox(width: 8),
          GestureDetector(
            onLongPress: _startRecording,
            onLongPressUp: _stopRecording,
            child: Container(
              width: 48,
              height: 48,
              decoration: BoxDecoration(
                color: Theme.of(context).primaryColor,
                shape: BoxShape.circle,
              ),
              child: Icon(
                _messageController.text.trim().isNotEmpty
                    ? Icons.send
                    : Icons.mic,
                color: Colors.white,
              ),
            ),
          ),
        ],
      ),
    );
  }

  void _showAttachmentOptions() {
    showModalBottomSheet(
      context: context,
      builder: (context) {
        return Container(
          height: 200,
          child: Column(
            children: [
              Padding(
                padding: EdgeInsets.all(16),
                child: Text(
                  'Send',
                  style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold),
                ),
              ),
              Expanded(
                child: GridView(
                  gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                    crossAxisCount: 3,
                    childAspectRatio: 1.2,
                  ),
                  children: [
                    _buildAttachmentOption(
                      Icons.camera_alt,
                      'Camera',
                      Colors.red,
                      () => _sendMedia('camera'),
                    ),
                    _buildAttachmentOption(
                      Icons.photo_library,
                      'Gallery',
                      Colors.purple,
                      () => _sendMedia('gallery'),
                    ),
                    _buildAttachmentOption(
                      Icons.insert_drive_file,
                      'Document',
                      Colors.blue,
                      () => _sendMedia('document'),
                    ),
                    _buildAttachmentOption(
                      Icons.location_on,
                      'Location',
                      Colors.green,
                      () => _sendLocation(),
                    ),
                    _buildAttachmentOption(
                      Icons.person,
                      'Contact',
                      Colors.orange,
                      () => _sendContact(),
                    ),
                  ],
                ),
              ),
            ],
          ),
        );
      },
    );
  }

  Widget _buildAttachmentOption(
    IconData icon,
    String label,
    Color color,
    VoidCallback onTap,
  ) {
    return GestureDetector(
      onTap: () {
        Navigator.pop(context);
        onTap();
      },
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Container(
            width: 48,
            height: 48,
            decoration: BoxDecoration(
              color: color,
              shape: BoxShape.circle,
            ),
            child: Icon(icon, color: Colors.white),
          ),
          SizedBox(height: 8),
          Text(label, style: TextStyle(fontSize: 12)),
        ],
      ),
    );
  }

  String _formatTime(DateTime dateTime) {
    DateTime now = DateTime.now();
    if (dateTime.day == now.day) {
      return '${dateTime.hour.toString().padLeft(2, '0')}:${dateTime.minute.toString().padLeft(2, '0')}';
    } else {
      return '${dateTime.day}/${dateTime.month}';
    }
  }

  void _scrollToBottom() {
    if (_scrollController.hasClients) {
      _scrollController.animateTo(
        _scrollController.position.maxScrollExtent,
        duration: Duration(milliseconds: 300),
        curve: Curves.easeOut,
      );
    }
  }

  void _sendMessage() {
    if (_messageController.text.trim().isNotEmpty) {
      setState(() {
        messages.add(Message(
          id: DateTime.now().millisecondsSinceEpoch.toString(),
          senderId: 'current_user',
          text: _messageController.text,
          timestamp: DateTime.now(),
          type: MessageType.text,
          isRead: false,
        ));
      });

      _messageController.clear();
      _scrollToBottom();
    }
  }

  void _sendMedia(String type) {
    setState(() {
      messages.add(Message(
        id: DateTime.now().millisecondsSinceEpoch.toString(),
        senderId: 'current_user',
        imageUrl: 'https://example.com/sent_media.jpg',
        timestamp: DateTime.now(),
        type: MessageType.image,
        isRead: false,
      ));
    });
    _scrollToBottom();
  }

  void _sendLocation() {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text('Location sharing functionality')),
    );
  }

  void _sendContact() {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text('Contact sharing functionality')),
    );
  }

  void _startRecording() {
    setState(() {
      isRecording = true;
    });
    // Start voice recording
  }

  void _stopRecording() {
    setState(() {
      isRecording = false;
    });
    // Stop recording and send voice message
    setState(() {
      messages.add(Message(
        id: DateTime.now().millisecondsSinceEpoch.toString(),
        senderId: 'current_user',
        timestamp: DateTime.now(),
        type: MessageType.voice,
        isRead: false,
      ));
    });
    _scrollToBottom();
  }

  void _startVideoCall() {
    Navigator.pushNamed(context, '/video-call', arguments: widget.username);
  }

  void _startVoiceCall() {
    Navigator.pushNamed(context, '/voice-call', arguments: widget.username);
  }

  void _navigateToProfile() {
    Navigator.pushNamed(context, '/user-profile', arguments: widget.username);
  }

  void _handleMenuAction(String action) {
    switch (action) {
      case 'view_profile':
        _navigateToProfile();
        break;
      case 'mute':
        ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(content: Text('Chat muted')),
        );
        break;
      case 'block':
        _showBlockDialog();
        break;
      case 'report':
        _showReportDialog();
        break;
      case 'clear_chat':
        _showClearChatDialog();
        break;
    }
  }

  void _showBlockDialog() {
    showDialog(
      context: context,
      builder: (context) => AlertDialog(
        title: Text('Block ${widget.username}?'),
        content: Text('Blocked contacts will not be able to send you messages.'),
        actions: [
          TextButton(
            onPressed: () => Navigator.pop(context),
            child: Text('Cancel'),
          ),
          TextButton(
            onPressed: () {
              Navigator.pop(context);
              ScaffoldMessenger.of(context).showSnackBar(
                SnackBar(content: Text('${widget.username} blocked')),
              );
            },
            child: Text('Block', style: TextStyle(color: Colors.red)),
          ),
        ],
      ),
    );
  }

  void _showReportDialog() {
    showDialog(
      context: context,
      builder: (context) => AlertDialog(
        title: Text('Report ${widget.username}?'),
        content: Text('This will report the user for inappropriate behavior.'),
        actions: [
          TextButton(
            onPressed: () => Navigator.pop(context),
            child: Text('Cancel'),
          ),
          TextButton(
            onPressed: () {
              Navigator.pop(context);
              ScaffoldMessenger.of(context).showSnackBar(
                SnackBar(content: Text('User reported')),
              );
            },
            child: Text('Report', style: TextStyle(color: Colors.red)),
          ),
        ],
      ),
    );
  }

  void _showClearChatDialog() {
    showDialog(
      context: context,
      builder: (context) => AlertDialog(
        title: Text('Clear Chat?'),
        content: Text('This will delete all messages in this chat.'),
        actions: [
          TextButton(
            onPressed: () => Navigator.pop(context),
            child: Text('Cancel'),
          ),
          TextButton(
            onPressed: () {
              Navigator.pop(context);
              setState(() {
                messages.clear();
              });
              ScaffoldMessenger.of(context).showSnackBar(
                SnackBar(content: Text('Chat cleared')),
              );
            },
            child: Text('Clear', style: TextStyle(color: Colors.red)),
          ),
        ],
      ),
    );
  }

  void _viewFullScreenImage(String imageUrl) {
    Navigator.push(
      context,
      MaterialPageRoute(
        builder: (context) => Scaffold(
          appBar: AppBar(
            backgroundColor: Colors.black,
            iconTheme: IconThemeData(color: Colors.white),
          ),
          backgroundColor: Colors.black,
          body: Center(
            child: InteractiveViewer(
              child: Image.network(
                imageUrl,
                fit: BoxFit.contain,
              ),
            ),
          ),
        ),
      ),
    );
  }

  void _playVoiceMessage(Message message) {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text('Playing voice message...')),
    );
  }

  @override
  void dispose() {
    _messageController.dispose();
    _scrollController.dispose();
    _animationController.dispose();
    super.dispose();
  }
}

enum MessageType {
  text,
  image,
  voice,
  video,
}

class Message {
  final String id;
  final String senderId;
  final String? text;
  final String? imageUrl;
  final DateTime timestamp;
  final MessageType type;
  bool isRead;

  Message({
    required this.id,
    required this.senderId,
    this.text,
    this.imageUrl,
    required this.timestamp,
    required this.type,
    required this.isRead,
  });
}